src/test/credentials/provider/ssoAccessTokenProvider.test.ts

```diff
@@ -0,0 +1,230 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as assert from 'assert'
+import * as AWS from 'aws-sdk'
+import * as sinon from 'sinon'
+import { SSOOIDC } from 'aws-sdk'
+import { DiskCache } from '../../../credentials/sso/diskCache'
+import { SsoAccessTokenProvider } from '../../../credentials/sso/ssoAccessTokenProvider'
+import { CreateTokenResponse, StartDeviceAuthorizationResponse } from 'aws-sdk/clients/ssooidc'
+import { SsoClientRegistration } from '../../../credentials/sso/ssoClientRegistration'
+import { SsoAccessToken } from '../../../credentials/sso/ssoAccessToken'
+
+describe('SsoAccessTokenProvider', () => {
+    let sandbox = sinon.createSandbox()
+
+    const ssoRegion = 'fakeRegion'
+    const ssoUrl = 'fakeUrl'
+    const ssoOidcClient = new SSOOIDC({ region: 'us-west-2' })
+    const cache = new DiskCache()
+    const sut = new SsoAccessTokenProvider(ssoRegion, ssoUrl, ssoOidcClient, cache)
+
+    const HOUR_IN_MS = 3600000
+
+    const validAccessToken = {
+        startUrl: ssoUrl,
+        region: ssoRegion,
+        accessToken: 'dummyAccessToken',
+        expiresAt: new Date(Date.now() + HOUR_IN_MS).toISOString(),
+    }
+
+    const fakeCreateTokenResponse: SSOOIDC.CreateTokenResponse = {
+        accessToken: 'dummyAccessToken',
+        expiresIn: 120,
+    }
+
+    const validRegistation: SsoClientRegistration = {
+        clientId: 'aString',
+        clientSecret: 'aString',
+        expiresAt: new Date(Date.now() + HOUR_IN_MS).toISOString(),
+    }
+
+    const validAuthorization: StartDeviceAuthorizationResponse = {
+        expiresIn: 120,
+        deviceCode: 'dummyCode',
+        userCode: 'dummyUserCode',
+        verificationUri: 'aUrl',
+        verificationUriComplete: 'aUrlComplete',
+        interval: 1,
+    }
+
+    function setUpStubCache(returnRegistration?: SsoClientRegistration, returnAccessToken?: SsoAccessToken) {
+        if (returnAccessToken) {
+            sandbox.stub(cache, 'loadAccessToken').returns(returnAccessToken)
+        } else {
+            sandbox.stub(cache, 'loadAccessToken').returns(undefined)
+        }
+
+        if (returnRegistration) {
+            sandbox.stub(cache, 'loadClientRegistration').returns(returnRegistration)
+        } else {
+            sandbox.stub(cache, 'loadClientRegistration').returns(undefined)
+        }
+
+        sandbox.stub(cache, 'saveClientRegistration').returns()
+    }
+
+    afterEach(async () => {
+        sandbox.restore()
+    })
+
+    describe('accessToken', () => {
+        it('should return a cached token', async () => {
+            sandbox.stub(cache, 'loadAccessToken').returns(validAccessToken)
+
+            const receivedToken = await sut.accessToken()
+
+            assert.strictEqual(receivedToken, validAccessToken)
+        })
+
+        it('should create a new access token with a cached client registration', async () => {
+            setUpStubCache(validRegistation)
```

Comment:
On the fence if I'd rather see this (or a chunk of the functionality) in a `beforeEach` instead of called at the start of each test.

Reply:
I'm interested in your thoughts.  I don't think tests need to follow DRY.  It makes sense to me to put testing framework actions and possibly environment details in the beforeEach.  Leaving all the test context in each test makes the file longer, but each test can stand on its own, making it more readable.